inc/cli/class-acm.php

```diff
@@ -0,0 +1,402 @@
+	 * : Allows for a more verbose output to the I/O
+	 *
+	 * [--dry_run]
+	 * : Test run for the command without changing any data
+	 *
+	 * @param array $args       Positional arguments.
+	 * @param array $assoc_args Associative arguments.
+	 */
+	public function acm( array $args, array $assoc_args ) {
+		$this->validate_inputs( $args, $assoc_args );
+
+		try {
+			$found_sites = 0;
+			$total_changed_sites = 0;
+			$progress_bar = null;
+			$paged = 1;
+			$sites_per_page = 100;
+
+			do {
+				$this->query_args['number'] = $sites_per_page;
+				$this->query_args['paged'] = $paged;
+				$this->query_args['offset'] = ( $paged - 1 ) * $sites_per_page;
+
+				$query = new \WP_Site_Query( $this->query_args );
+
+				if ( ! $progress_bar ) {
+					$found_sites = $query->found_sites;
+					$progress_bar = WP_CLI\Utils\make_progress_bar( 'Progress: ', $found_sites );
+				}
+
+				if ( empty( $query->sites ) ) {
+					break;
+				}
+
+				for ( $i = 0; $i < count( $query->sites ); $i++ ) {
+					if ( $i % 5 === 0 ) {
+						// Sleep for 1 second every 5 sites to avoid rate limiting.
+						sleep( 1 );
+					}
+
+					$site_id = $query->sites[ $i ];
+
+					if ( $this->verbose ) {
+						WP_CLI::log( sprintf( 'Running HM ACM %1$s command for site %2$d', $this->action, $site_id ) );
+					}
+
+					switch_to_blog( $site_id );
+
+					$result = false;
+
+					if ( $this->dry_run ) {
+						$result = true; // Dry run mode, so just assume everything goes well.
+					} else {
+						switch ( $this->action ) {
+							case 'create-cert':
+								$result = $this->create_cert( $site_id, $assoc_args );
+								break;
+							case 'verify-cert':
+								$result = $this->verify_cert( $site_id );
+								break;
+							case 'delete-cert':
+								$result = $this->delete_cert( $site_id );
+								break;
+							case 'create-cloudfront':
+								$result = $this->create_cloudfront( $site_id );
+								break;
+							default:
+								break;
+						}
+					}
+
+					if ( $result ) {
+						$total_changed_sites++;
+					}
+
+					if ( $this->verbose && $result ) {
+						WP_CLI::success( sprintf( 'Site %1$d HM ACM %2$s command result was successful.', $site_id, $this->action ) );
+					}
+
+					if ( $this->verbose && ! $result ) {
+						WP_CLI::warning( sprintf( 'Site %1$d HM ACM %2$s command result failed.', $site_id, $this->action ) );
+					}
+
+					restore_current_blog();
+
+					if ( $this->verbose === false ) {
+						$progress_bar->tick();
+					}
+				}
+
+				$paged++;
+			} while ( ! empty( $query->sites ) );
+
+			if ( $this->verbose === false ) {
+				$progress_bar->finish();
+			}
+
+			if ( empty( $total_changed_sites ) ) {
+				WP_CLI::error( 'No sites were affected by the HM ACM command.' );
+			} else {
+				WP_CLI::success( sprintf( 'Successfully executed HM ACM command for %1$d/%2$d sites.', $total_changed_sites, $found_sites ) );
+			}
+
+			switch ( $this->action ) {
+				case 'create-cert':
+					$this->output_cert_records_csv();
```

Comment:
I've looked into this and while I've got it working via `wp xyz --format=csv | tee myfile.csv`, the progress bar no longer works and it's been useful to have that so that one can get quick feedback on whether the command is still executing as expected or has got stuck somewhere. Ideally would like to retain this.

Another slight downside is that the resulting csv would contain some other output even if not in verbose mode but that can be removed manually when file is opened.

Any thoughts on being able to do both output to terminal and file while also retaining the progress bar ?

Reply:
Hmm I thought the progress bar output to stderr, but maybe that's a WP CLI thing where it doesnt work while being piped. I think as long as you've considered it and we're doing it for a reason then let's not hold this up.